invokeai/backend/util/devices.py

```diff
@@ -1,89 +1,120 @@
-from __future__ import annotations
-
-from contextlib import nullcontext
-from typing import Literal, Optional, Union
+from typing import Dict, Literal, Optional, Union
+from warnings import warn
 
 import torch
-from torch import autocast
 
-from invokeai.app.services.config.config_default import PRECISION, get_config
+from invokeai.app.services.config.config_default import InvokeAIAppConfig, get_config
 
+# legacy APIs
+TorchPrecisionNames = Literal["float32", "float16", "bfloat16"]
 CPU_DEVICE = torch.device("cpu")
 CUDA_DEVICE = torch.device("cuda")
 MPS_DEVICE = torch.device("mps")
 
 
+def choose_precision(device: torch.device) -> TorchPrecisionNames:
+    """Return the string representation of the recommended torch device."""
+    warn(
+        "choose_precision() is deprecated. Use TorchDevice.choose_torch_dtype() instead.",
+        DeprecationWarning,
+        stacklevel=2,
+    )
```

Comment:
Suggest the `Deprecated` package as a cleaner way to do this. It's already a dependency. https://deprecated.readthedocs.io/en/latest/tutorial.html

Reply:
Thanks for this! It is much cleaner. I noticed that using the decorator makes mypy redline the function with `Untyped decorator makes function "choose_precision" untyped [misc]`. I put in a `# type: ignore`, but should I be concerned?

Reply:
My editor correctly shows the type signature of the function with or without the decorator - I think it's fine to shush mypy here.